tfhe/src/high_level_api/integers/unsigned/base.rs

```diff
@@ -477,6 +479,126 @@ where
             }
         })
     }
+
+    /// `match` an input value to an output value
+    ///
+    /// - Input values are not required to span all possible values that
+    ///  `self` could hold. And the output type can be different.
+    ///
+    /// Returns a FheBool that encrypts `true` if the input `self`
+    /// matched one of the possible inputs
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::prelude::*;
+    /// use tfhe::{
+    ///     generate_keys, set_server_key, ConfigBuilder, FheBool, FheUint16, FheUint8, MatchValues,
+    /// };
+    ///
+    /// let (client_key, server_key) = generate_keys(ConfigBuilder::default());
+    /// set_server_key(server_key);
+    ///
+    /// let a = FheUint16::encrypt(17u16, &client_key);
+    ///
+    /// let match_values = MatchValues::new(vec![
+    ///     (0u16, 3u16),
+    ///     (1u16, 3u16),
+    ///     (2u16, 3u16),
+    ///     (17u16, 25u16),
+    /// ])
+    /// .unwrap();
+    /// let (result, matched): (FheUint8, _) = a.match_value(&match_values);
+    ///
+    /// let matched = matched.decrypt(&client_key);
+    /// assert_eq!(matched, true);
+    ///
+    /// let decrypted: u16 = result.decrypt(&client_key);
+    /// assert_eq!(decrypted, 25u16)
+    /// ```
+    pub fn match_value<Clear, OutId>(
+        &self,
+        matches: &MatchValues<Clear>,
+    ) -> (FheUint<OutId>, FheBool)
+    where
+        Clear: UnsignedInteger + DecomposableInto<u64> + CastInto<usize>,
+        OutId: FheUintId,
+    {
+        global_state::with_internal_keys(|key| match key {
+            InternalServerKey::Cpu(cpu_key) => {
+                let (result, matched) = cpu_key
+                    .pbs_key()
+                    .match_value_parallelized(&self.ciphertext.on_cpu(), matches);
+                let result = cpu_key
+                    .pbs_key()
+                    .cast_to_unsigned(result, OutId::num_blocks(cpu_key.message_modulus()));
```

Comment:
Why do the cast in this operation?
A user could do it after, right?

Reply:
The number of blocks of the output of the interger function depends on the max possible output value that is in the MatchValue map allowing to map 16blocks value to values with less or more blocks

Since the HLAPI API types are not as dynamic, we have to let the user cast to the wanted type